src/activity/activity_heartbeat_manager.rs

```diff
@@ -1,86 +1,119 @@
+            Some(handle) => {
+                handle.heartbeat_tx.send(heartbeat.details);
+            }
+            None => {
+                let heartbeat_timeout: time::Duration = heartbeat
+                    .heartbeat_timeout
+                    .ok_or(ActivityHeartbeatError::HeartbeatTimeoutNotSet)?
+                    .try_into()
+                    .or(Err(ActivityHeartbeatError::InvalidHeartbeatTimeout))?;
+                let delay = heartbeat_timeout.div(2);
+                let (heartbeat_tx, heartbeat_rx) = channel(heartbeat.details);
+                let processor = ActivityHeartbeatProcessor {
+                    delay,
+                    heartbeat_rx,
+                    shutdown_rx: self.shutdown_rx.clone(),
+                };
+                let join_handle = tokio::spawn(processor.run());
+                let handle = ActivityHeartbeatProcessorHandle {
+                    heartbeat_tx,
+                    join_handle,
+                };
+                self.heartbeat_processors
+                    .insert(heartbeat.task_token, handle);
+            }
+        }
         Ok(())
     }
 
     pub fn shutdown(&self) {
         self.shutdown_tx.send(true);
+        // TODO join on join handles
     }
+}
 
-    fn processor(heartbeat_rx: Receiver<ActivityHeartbeatTask>, shutdown_rx: Receiver<bool>) {
-        // TODO this function should:
-        // 1. consume heartbeat tasks
-        // 2. join the futures and for any future that becomes ready:
-        // - send heartbeat details to the server (if any)
-        // - clear details from the local cache
-        // - schedule next task in case if heartbeat has been sent, otherwise clear task token from scheduled_heartbeats
-        unimplemented!()
+impl ActivityHeartbeatProcessor {
+    async fn run(mut self) {
+        // TODO borrow and send details available in the heartbeat_rx once.
+        loop {
+            sleep(self.delay).await;
+            let stop = select! {
+                _ = self.shutdown_rx.changed() => {
+                    // Shutting down core, need to break the loop. Previous details has been sent,
+                    // so there is nothing else to do.
+                    true
+                }
+                _ = sleep(self.delay) => {
+                    // Timed out while waiting for the next heartbeat.
+                    // We waited 2 * delay in total, where delay is 1/2 of the activity heartbeat timeout.
+                    // This means that activity has either timed out or completed by now.
+                    true
+                }
+                _ = self.heartbeat_rx.changed() => {
+                    // Received new heartbeat details.
+                    let details = self.heartbeat_rx.borrow();
+                    // TODO send details right away since enough time has passed since previous call
+                    false
+                }
+            };
+            if stop {
+                break;
```

Comment:
Rather than returning a bool and then `break`ing here you could just do it inside the select. Ultimately a bit less code.

Reply:
Done, for some reason I thought that break would break out of the select initially.